src/coord/src/coord.rs

```diff
@@ -839,29 +846,67 @@ impl Coordinator {
                             );
                         }
                         if index_state.since != compaction_frontier {
+                            // Mark previous since for eviction.
+                            let mut updates: Vec<_> = index_state
+                                .since
+                                .elements()
+                                .iter()
+                                .map(|time| (*time, -1))
+                                .collect();
                             index_state.advance_since(&compaction_frontier);
-                            self.since_updates
-                                .push((name.clone(), index_state.since.clone()));
+                            // Mark new since for insertion.
+                            for time in index_state.since.elements() {
+                                updates.push((*time, 1));
+                            }
+                            // We always want `compaction_frontiers` to have exactly one timestamp tracking
+                            // the index's since, so send both the eviction and insertion in the same
+                            // update batch.
+                            self.update_compaction_frontier(*name, updates);
                         }
                     }
                 }
             }
         }
     }
 
+    /// Updates the tracked compaction frontier for an index.
+    ///
+    /// If the antichain progressed, it is added to the pending compactions.
```

Comment:
I am getting the following error in this branch *at random points* with the attached SQL test case:

[5849.zip](https://github.com/MaterializeInc/materialize/files/6038142/5849.zip)

```
      1 ERROR:  Dataflow temp-view-t105954 requested as_of (Antichain { elements: [1614190467993] }) not >= since (Antichain { elements: [1614190468003] })
```

I will try to simplify the test case tomorrow, but it seems that the database needs to get into a particular state before the error starts being thrown. When that happens, we get the following output for the second identical SELECT query within the same transaction:

```
BEGIN;
BEGIN
SELECT COUNT(*) FROM t1 /* rqg_query_id : 36634624 */;
 count
-------
  2449
(1 row)

SELECT COUNT(*) FROM t1 /* rqg_query_id : 299958272 */;
ERROR:  Dataflow temp-view-t152722 requested as_of (Antichain { elements: [1614191573993] }) not >= since (Antichain { elements: [1614191574026
```

Reply:
I think this is happening because transactions don't prevent you from later querying from sources not from the initial time domain. Working on fixing this then could use a retest (not yet).

Reply:
Ok I'm second guessing myself now because that would only happen if you used multiple databases or schemas, which I find doubtful? Can you post all of the text between a failure and the previous `COMMIT` or `ROLLBACK`?

Reply:
What seems to be common between those scenarios is that the previous transaction was a DDL (sometimes against a view that is not then used). Here is typical sequence:

```
COMMIT
CREATE OR REPLACE  VIEW v1 AS SELECT COUNT(*) FROM t1 /* rqg_query_id : 436273152 */;
CREATE VIEW
BEGIN;
BEGIN
SELECT COUNT(*) FROM t1 /* rqg_query_id : 1155137536 */;
 count 
-------
  9075
(1 row)

SELECT COUNT(*) FROM v1 /* rqg_query_id : 1727987712 */;
 count
-------
     1
(1 row)

SELECT COUNT(*) FROM t1 /* rqg_query_id : 2256601088 */;
 count
-------
  9075
(1 row)

SELECT COUNT(*) FROM t1 /* rqg_query_id : 2772041728 */;
ERROR:  Dataflow temp-view-t36574 requested as_of (Antichain...
```

Reply:
Unfortunately it is not possible to simplify the test case further in any meaningful way. It seems that a certain number of work needs to have happened against the database for the issue to start showing up. Maybe it sub-millisecond logical-compaction-windows are allowed the issue will become more reproducible. However, currently attempting to set logical-compaction-window smaller than 1ms results in a "divide by zero" panic.

Please let me know if there is anything else I can help with regarding this particular situation.

Reply:
I'm able to reproduce this easily locally. Thanks for the initial find.

Reply:
> I think this is happening because transactions don't prevent you from later querying from sources not from the initial time domain. Working on fixing this then could use a retest (not yet).

Yes I meant to ask about that -- are the sources that participate in the first statement in a transaction special in any way? Is it the case that compaction in the entire Mz instance needs to be stopped until the transaction completes? (similar to how other databases will not purge their write-ahead-logs beyond the longest running transaction) .

Reply:
Nevermind, I now see that you are collecting the schemas of the database objects of the first statement rather than the objects themselves. So please ignore previous question.